protocols/mdns/src/dns.rs

```diff
@@ -80,68 +101,70 @@ pub fn build_query() -> Vec<u8> {
     out
 }
 
-/// Builds the response to the DNS query.
+/// Builds the response to an address discovery DNS query.
 ///
 /// If there are more than 2^16-1 addresses, ignores the rest.
 pub fn build_query_response(
     id: u16,
     peer_id: PeerId,
     addresses: impl ExactSizeIterator<Item = Multiaddr>,
     ttl: Duration,
-) -> Result<Vec<u8>, MdnsResponseError> {
+) -> Vec<MdnsPacket> {
     // Convert the TTL into seconds.
     let ttl = duration_to_secs(ttl);
 
     // Add a limit to 2^16-1 addresses, as the protocol limits to this number.
-    let addresses = addresses.take(65535);
-
-    // This capacity was determined empirically and is a reasonable upper limit.
-    let mut out = Vec::with_capacity(320);
+    let mut addresses = addresses.take(65535);
 
-    append_u16(&mut out, id);
-    // 0x84 flag for an answer.
-    append_u16(&mut out, 0x8400);
-    // Number of questions, answers, authorities, additionals.
-    append_u16(&mut out, 0x0);
-    append_u16(&mut out, 0x1);
-    append_u16(&mut out, 0x0);
-    append_u16(&mut out, addresses.len() as u16);
+    let peer_id_bytes = encode_peer_id(&peer_id);
+    debug_assert!(peer_id_bytes.len() <= 0xffff);
 
-    // Our single answer.
-    // The name.
-    append_qname(&mut out, SERVICE_NAME);
+    let mut packets = Vec::new();
+
+    'next_packet: loop {
```

Comment:
It took me a bit to understand this loop. The following control flow would be easier for me to understand. What do you think?

```diff
diff --git a/protocols/mdns/src/dns.rs b/protocols/mdns/src/dns.rs
index efb5a11a..fcc62fc2 100644
--- a/protocols/mdns/src/dns.rs
+++ b/protocols/mdns/src/dns.rs
@@ -120,38 +120,23 @@ pub fn build_query_response(
     debug_assert!(peer_id_bytes.len() <= 0xffff);
 
     let mut packets = Vec::new();
-
-    'next_packet: loop {
-        let mut records = Vec::with_capacity(addresses.len() * MAX_TXT_RECORD_SIZE);
-
-        // Try to pack as many records into an MDNS response packet
-        // as are known to fit comfortably.
-        for _ in 0 .. MAX_RECORDS_PER_PACKET {
-            if let Some(addr) = addresses.next() {
-                let txt_to_send = format!("dnsaddr={}/p2p/{}", addr.to_string(), peer_id.to_base58());
-                let mut txt_record = Vec::with_capacity(txt_to_send.len());
-                match append_txt_record(&mut txt_record, &peer_id_bytes, ttl, &txt_to_send) {
-                    Ok(()) => {
-                        records.push(txt_record);
-                        continue
-                    }
-                    Err(e) => {
-                        log::warn!("Excluding address {} from response: {:?}", addr, e);
-                    }
-                }
+    let mut records = Vec::with_capacity(addresses.len() * MAX_TXT_RECORD_SIZE);
+
+    while let Some(addr) = addresses.next() {
+        let txt_to_send = format!("dnsaddr={}/p2p/{}", addr.to_string(), peer_id.to_base58());
+        let mut txt_record = Vec::with_capacity(txt_to_send.len());
+        match append_txt_record(&mut txt_record, &peer_id_bytes, ttl, &txt_to_send) {
+            Ok(()) => {
+                records.push(txt_record);
+                continue
             }
-
-            // Push the final packet, if there are still unpacked records.
-            if !records.is_empty() {
-                packets.push(query_response_packet(id, &peer_id_bytes, &records, ttl));
+            Err(e) => {
+                log::warn!("Excluding address {} from response: {:?}", addr, e);
             }
-
-            break 'next_packet
         }
-
-        // Reached MAX_RECORDS_PER_PACKET. Finish up the current packet and
-        // begin a new one.
-        packets.push(query_response_packet(id, &peer_id_bytes, &records, ttl));
+        if records.len() == MAX_RECORDS_PER_PACKET || addresses.len() == 0 {
+            packets.push(query_response_packet(id, &peer_id_bytes, &records, ttl));
+        }
     }
```

Reply:
What do you think about https://github.com/libp2p/rust-libp2p/pull/1877/commits/397b25026db0ea6103790c2b0873c4b36c0d5d40? With your suggestion as-is, I think the `continue` in the `Ok` branch would effectively render `MAX_RECORDS_PER_PEER` ineffective.

Reply:
Thanks, looks good to me. Indeed I missed the `continue`.